### ISSUE TITLE
`igraph_disjoint_union()` should preserve attributes

### ISSUE THREAD

**Issue by szhorvat** [igraph/igraph#2492]
**What is the feature or improvement you would like to see?**

`igraph_disjoint_union()` should preserve vertex and edge attributes. This is feasible since there is a one-to-one mapping between the edges and vertices of the input pair and the output graph.

**Use cases for the feature**

Disjoint union is one of the most common ways to build graphs.

**References**
—

**Comment by ntamas** [comment]
No, we don't have an API for that yet. I was thinking that we could solve this with `permute_vertices()` and `permute_edges()` but apparently these currently assume that the target graph has no vertex/edge attributes yet.

Tbh the `copy()` operation with its three boolean attributes always felt a bit out of place for me in the attribute handler table. Maybe we could deprecate it and add `copy_graph_attributes()`, `copy_vertex_attributes()` and `copy_edge_attributes()` instead? These functions could take a mapping from the old vertex/edge indices to the new ones to achieve what you need here. What I don't like about this is that 1) `copy_vertex_attributes()` and `copy_edge_attributes()` would essentially duplicate the functionality of `permute_vertices()` and `permute_edges()`, and 2) constructing a vector of numbers just for the sake of mapping a _range_ of old vertex IDs to a _range_ of new vertex IDs seems unnecessary.

**Comment by szhorvat** [comment]
Would it make sense to have some sort of interface for concatenating attribute lists, to complement the existing one that permutes them?

Python stores attributes in a Python list. There are efficient ways to concatenate these. So maybe it's a good idea to try to avoid handling attribute values one by one (vertex by vertex, edge by edge) in C code.

This is difficult because we need to find a good balance between the flexibility to manipulate attributes of arbitrary types in arbitrary ways, and the efficiency of offloading this work to the high-level interface that can best handle its own datatypes.
